Browser support and JS Guidelines Cleanup (#12)
inrupt/public-documentation: coding-conventions/javascript-coding-standards.md

# JavaScript-Specific Guidelines

## Browser Support

- Test and target compilation for the latest two versions of the most popular browsers. The current list is:

>   - Desktop:

Reviewer: As with the Mobile browsers, do we want to be explicit on the browser/OS combinations:
- Chrome / macOS, Ubuntu, and Windows
- Edge (Chromium) / macOS and Windows
- Firefox / macOS, Ubuntu, and Windows
- Safari / macOS

>     -  Edge

Reviewer: I would explicitly say "Edge (Chromium)"

Author: Wouldn't "latest two versions" cover chromium?

Reviewer: Edge (edgeHTML) and Edge (Chromium) are 2 separate products, and it seems as if Microsoft treat the former still as a live product. Probably doesn't make any much of a difference (who uses the edgeHTML version??) but... Your call.

Author: Fair enough, I'll call out Chromium.

    - Medium: 768x1024
    - Large: 1366x768
    - X-Large: 1600x900

## Organize code conceptually in packages

- Use a folder hierarchy to represent packages of related code.
- Dependencies are usually tree-shaped.
  - Other graph shapes might indicate an architectural problem.
- One index file with non-default ES6 exports exposes the public API.

## Coding style

### Defensive programming

- Validate your assumptions (at the moment you make them):
  - For typing, this is possible through TypeScript.
  - Methods throw an error when called with invalid arguments.

### Asynchronicity

- Prefer async/await over explicit Promises, and prefer those over
  callbacks.
  - Promise wrappers for native Node functions exist.
- Use callbacks if needed for performance reasons.
  - Especially for streams.

## Testing

- Ensure all code is tested with unit tests, or has a comment explaining why there's no test and is
  [skipped for the coverage report](https://github.com/gotwarlost/istanbul/blob/master/ignoring-code-for-coverage.md)
  - Classes have corresponding unit tests.
  - Replace dependencies by mocks with validated assumptions.
  - High coverage is important, but not a goal in itself.

>   - Not only ensure that code is used, but that its effects are tested.

Reviewer 2: ? so that the form is ensure not only this but that.

```suggestion
Ensure not only that the code is used, but that is effects are tested.
```

>   - Ensure not only that the code is used, but that is effects are tested.

Reviewer 3: ```suggestion
  - Ensure not only that the code is used, but that its effects are tested.
```

- Write integration tests for groups of classes or functions.

>   - Since each has one specific responsibility, integration tests check how those responsibilities

Reviewer 2: By "each" do you mean, each class/function?

    interact. 

## Tools and Quality Practices

1. Add an `.nvmrc` file to your repository, and [set up your shell](https://github.com/nvm-sh/nvm#deeper-shell-integration)
  to read it and switch to the correct Node/npm version automatically.
2. Use Node 12.
3. Use npm.
4. Use TypeScript for larger projects; ES6 JavaScript (through Babel) is acceptable for apps.

>   1. Put source code on the `src` folder.

Reviewer 3: ```suggestion
  1. Put source code in the `src` folder.
```

  2. Put transpiled code in the `dist` folder.

>   3. Add non-default export to an index file, ensuring that tree-shaking is possible with the `sideEffects` flag in `package.json`.

Reviewer 3: Maybe an example snippet  - i.e. I have no idea if it's:
```
{
  sideEffects: true  // Or is it false!
}
```
Or just link to the WebPack docs: https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free
And is this only relevant for WebPack ('cos we use both Parcel and Rollup now too!)...?

> 5. Marking everything API that\'s intended for internal use but is externally accessible as `@internal` in its TSDoc, and

Reviewer 3: ```suggestion
5. Annotate everything in the API that\'s intended for internal use, but is externally accessible, as `@internal` in its TSDoc, and
```

  [prevent TypeScript from emitting their declarations](https://www.typescriptlang.org/v2/en/tsconfig#stripInternal)
6. Only use the `any` type as a last resort - which in practice basically means *never*. If you *do* use
  it, accompany it with a comment explaining why you chose to use it anyway. Keep in mind that some

>   native methods return `any`, like `JSON.parse` -  those to a known type as soon as possible, preferably

Reviewer 2: missing a verbe before "those"?

  with a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards).
7. Write tests with Jest.

>   1. Put tests next to src files. In libraries, this is typically done by having a `foo.ts` and `foo.test.ts`

Reviewer 3: ```suggestion
  1. Put tests next to source files. In libraries, this is typically done by having a `foo.ts` and `foo.test.ts`
```

    file; in apps, this is typically done by
    using folders (`foo/index.jsx`, `foo/index.test.jsx`, `foo/styles.js`).
8. Use ESLint.
  1. On all code, including tests.
  2. Use the [Inrupt standard eslint configs](https://github.com/inrupt/javascript-style-configs).
9. Use Prettier on all code, including tests. (This will be done by the linter.)
10. Collect code coverage statistics with Jest.
  1. Strive for 100% branch coverage.
  2. Coverage does not tell the full story; for complex projects, consider TestCafe as well.
11. Set up commit hooks with Husky to lint, compile, and test before pushing.
12. Set up GitHub actions to run on Node 12 and LTS.
13. One repository is one npm module by default.
  1. If more complex reuse becomes a necessity, consult with others before converting to a multi-package repo with Lerna.
14. When building applications, use the Next.js framework. For documenting components, use Storybook. For libraries, use rollup.

>   1. Generate minified and non-minified versions.

Reviewer 3: Is it worth showing simple guidance (or a `package.json` snippet, or whatever) to show how ('cos I don't know!)...?

>   2. Ensure that all generated versions export a mapping file.

Reviewer 3: Is it worth showing simple guidance (or a `package.json` snippet, or whatever) to show how ('cos I don't know!)...?

> 15. When including logging, use loglevel.

Reviewer 3: ```suggestion
15. When including logging, use [loglevel](https://www.npmjs.com/package/loglevel).
```
